include/revng/Support/GraphAlgorithms.h

```diff
@@ -196,3 +201,215 @@ auto exitless_scc_range(NodeTy Entry) {
+                                       detail::GT<NodeRef>,
+                                       detail::GT<llvm::Inverse<NodeRef>>>;
+
+  using Iterator = VariedIterator<typename GraphTrait::nodes_iterator,
+                                  typename NodeTrait::ChildIteratorType>;
+
+public:
+  ValueType Value;
+
+public:
+  GraphNodeOrRootHelper(GraphType Graph) : Value(Graph) {}
+  GraphNodeOrRootHelper(llvm::Inverse<GraphType> Inv) : Value(Inv.Graph) {}
+  GraphNodeOrRootHelper(NodeRef Node) : Value(std::move(Node)) {}
+  GraphNodeOrRootHelper(std::nullptr_t) : Value((GraphType) nullptr) {}
+  GraphNodeOrRootHelper(EquivalentPair P) : Value((GraphType) nullptr) {
+    if (P.first != nullptr)
+      Value = P.first;
+    else
+      Value = P.second;
+  }
+
+public:
+  Iterator child_begin() {
+    auto Visitor = OverloadedLambda{
+      [](GraphType Graph) { return Iterator(GraphTrait::nodes_begin(Graph)); },
+      [](NodeRef Node) { return Iterator(NodeTrait::child_begin(Node)); }
+    };
+    return std::visit(Visitor, Value);
+  }
+  Iterator child_end() {
+    auto Visitor = OverloadedLambda{
+      [](GraphType Graph) { return Iterator(GraphTrait::nodes_end(Graph)); },
+      [](NodeRef Node) { return Iterator(NodeTrait::child_end(Node)); }
+    };
+    return std::visit(Visitor, Value);
+  }
+
+public:
+  NodeRef *get() {
+    auto Visitor = OverloadedLambda{
+      [](GraphType) -> NodeRef * { return nullptr; },
+      [](NodeRef &Node) -> NodeRef * { return &Node; }
+    };
+    return std::visit(Visitor, Value);
+  }
+  const NodeRef *get() const {
+    auto Visitor = OverloadedLambda{
+      [](GraphType) -> const NodeRef * { return nullptr; },
+      [](const NodeRef &Node) -> const NodeRef * { return &Node; }
+    };
+    return std::visit(Visitor, Value);
+  }
+
+  operator bool() const { return get(); }
+  operator std::optional<NodeRef>() const {
+    if (auto Result = get())
+      return *Result;
+    else
+      return std::nullopt;
+  }
+
+  NodeRef &operator*() {
+    revng_assert(get() != nullptr);
+    return *get();
+  }
+  const NodeRef &operator*() const {
+    revng_assert(get() != nullptr);
+    return *get();
+  }
+
+  NodeRef *operator->() {
+    revng_assert(get() != nullptr);
+    return get();
+  }
+  const NodeRef *operator->() const {
+    revng_assert(get() != nullptr);
+    return get();
+  }
+
+public:
+  operator EquivalentPair() {
+    auto Visitor = OverloadedLambda{
+      [](GraphType Graph) { return std::make_pair(Graph, NodeRef{}); },
+      [](NodeRef Node) { return std::make_pair((GraphType) nullptr, Node); }
+    };
+    return std::visit(Visitor, Value);
+  }
+  operator EquivalentPair() const {
+    auto Visitor = OverloadedLambda{
+      [](GraphType Graph) { return std::make_pair(Graph, NodeRef{}); },
+      [](NodeRef Node) { return std::make_pair((GraphType) nullptr, Node); }
+    };
+    return std::visit(Visitor, Value);
+  }
```

Comment:
Implicit casts are very evil. Why do we need this?

Reply:
Actually, these can (and should) be explicit. As to "why", see the other thread.

Reply:
Closing, let's continue in the other thread.